### PR TITLE
Implement the proc-macro version.

### PR DESCRIPTION
**Pull request by richard-uk1**
Generate a widget struct for the enum Data and implement Widget<Data> for it.

### CODE REVIEW

**Comment by richard-uk1** [comment]
It's not complete yet because I don't know how to do adding/removing widgets dynamically (needed in the `update` method)

**Comment by luleyleo** [comment]
This looks like a good start!

Yea, `update` and `lifecycle` need a little bit of extra functionality.

What we need to do:
- in our widget struct add  `discriminant: Option<Discriminant<#name>>` and set it to `None` initially as we have no data yet.
- in `lifecycle` update `self.discriminant` to `Some(mem::discriminant(data))` when we receive `LifeCycle::WidgetAdded`.
- in `update` compare `self.discriminant` with `mem::discriminant(data)`
  - call `ctx.children_changed()` if they are different and update `self.discriminant`.
  - call the right update method (`update_match`) if they are the same.

And probably also:

- in `event`, `layout` and `paint` compare `self.discriminant` with `mem::discriminant(data)`
  - skip the rest if they are different
  - call the right method (`*_match`) if they are the same.

The last one will likely be necessary because it's possible that the variant changes during `event` which could cause a branch of widgets that have not yet received `WidgetAdded` to receive events. But I'll have to think about that a bit more.

So `update` would look pretty much like this:
```rust
if self.discriminant != Some(mem::discriminant(data)) {
    ctx.children_changed();
    self.discriminant = Some(mem::discriminant(data));
} else {
    match data {
        #(#update_match)*
    }
}
```

**Comment by richard-uk1** [comment]
I've implemented this, just playing with it now to see if it works.

One way of checking exhaustiveness would be to generate two structs:

 1. A `MatcherBuilder` where all the fields are options (with builder methods to set them)
 2. A `Matcher` struct that doesn't have any `Option`s.

You'd then have a `build` method on the builder that gives you the matcher if everything is set, or panics/returns an error if not.

**Comment by richard-uk1** [comment]
I've added your suggestions. One think I did differently is matching on `(old_data, data)` rather than using the `discriminant` in `update`, because I need to pass the old data into the child function as well.

**Comment by richard-uk1** [comment]
I'm building an app that will use this to test it: https://github.com/derekdreery/yatzy

File: match-derive/src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,179 @@
...
-    let _ = input;
-    unimplemented!()
+mod parse;
+
+use parse::MatcherDerive;
+
+#[proc_macro_derive(Matcher, attributes(matcher))]
+pub fn derive(input: TokenStream) -> TokenStream {
+    // TODO when we generate a name that isn't a valid ident or is a keyword, generate a different
+    // name rather than panicking.
```
Comment (luleyleo) [inline match-derive/src/lib.rs:12]: I think it should cause a compilation error, but not do any name picking magic.

**Review by luleyleo — APPROVED** [review]
That's cool!

Regarding the exhaustiveness:
My idea would be that by default it gets checked lazily during `LifeCycle::WidgetAdded` where it would log a waring in the case a variant is missing and no default is provided.
I addition to that we would provide `or_else(widget)`, `or_empty()` and possibly `check()` methods where the first two can be used to specify a default widget and `check` could be use to explicitly request a `panic` in place if the matcher is non-exhaustive.

So the remaining major todo's would be:
- default variants
- checking exhaustiveness
- generics
- tuple variants
- documentation of the derive
- generating documentation for the `...Matcher` widget and its builders
- ensure consistency with the guarantees druid offers to widgets
  (e.g. `LifeCycle::WidgetAdded` as first event)
- bullet proving error messages
- nitpicking about implementation details
- getting this into druid

I think I'll merge this and play with it a bit.

File: match-derive/src/lib.rs (hunk 1 above)
Comment (richard-uk1) [reply match-derive/src/lib.rs:12]: There is prior art for this in `wasm-bindgen`, where they append a `_` (underscore, e.g. `match_`) if there is a name clash. However there they have to generate equivalent rust names for wepapi interfaces, which is not exactly the same situation as us, and so we shouldn't necessarily do what they do.

**Comment by richard-uk1** [comment]
I have to say this would be a really nice thing to get into druid, because it solves the sum (`enum`) data problem (product (`struct`) data is easier to handle because it doesn't involve replacing the widget tree, and so can be done perfectly well in user code).

**Comment by richard-uk1** [comment]
Also note that currently even though I generate code to store the discriminant, it isn't actually used anywhere and could be removed. I left it in since from what you have said it will be needed in the event handling function.

**Pull request merged**